crates/wysiwyg/src/composer_model.rs

```diff
@@ -298,6 +310,95 @@ where
         }
     }
 
+    fn replace_multiple_nodes(
+        &mut self,
+        range: MultipleNodesRange,
+        new_text: &[C],
+    ) {
+        let mut to_delete = Vec::new();
+        for loc in range.into_iter() {
+            match loc.location_type {
+                RangeLocationType::Start => {
+                    let mut node =
+                        self.state.dom.lookup_node_mut(loc.node_handle);
+                    match &mut node {
+                        DomNode::Container(_) => {
+                            panic!("Start node must be text!")
+                        }
+                        DomNode::Text(node) => {
+                            let old_data = node.data();
+                            let mut new_data =
+                                old_data[..loc.start_offset].to_vec();
+                            new_data.extend_from_slice(new_text);
+                            node.set_data(new_data);
+                        }
+                    }
+                }
+                RangeLocationType::Middle => {
+                    let node =
+                        self.state.dom.lookup_node(loc.node_handle.clone());
+                    match node {
+                        DomNode::Container(_) => {
+                            // Ignore containers for now
+                        }
+                        DomNode::Text(_) => {
+                            to_delete.push(loc.node_handle);
+                        }
+                    }
+                }
+                RangeLocationType::End => {
+                    let mut node =
+                        self.state.dom.lookup_node_mut(loc.node_handle.clone());
+                    match &mut node {
+                        DomNode::Container(_) => {
+                            panic!("End node must be text!")
+                        }
+                        DomNode::Text(node) => {
+                            // TODO: maybe no need for RangeLocation here?
+                            let data = node.data();
+                            if loc.end_offset < data.len() {
+                                node.set_data(
+                                    node.data()[loc.end_offset..].to_vec(),
+                                );
+                            } else {
+                                to_delete.push(loc.node_handle);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Delete in reverse order to avoid invalidating handles
+        to_delete.reverse();
+
+        while !to_delete.is_empty() {
```

Comment:
The logic of `replace_multiple_nodes` and especially this loop is quite clever, but it took me a while to understand how this worked. Maybe we could add some comments explaining how the function works?

Reply:
Yes, definitely. I've just refactored it heavily in my next PR, so if I promise to it do after that, can we merge this?

Reply:
Sure, that's fine 👍